iroh-net/src/magic_endpoint.rs

```diff
@@ -455,122 +456,141 @@ mod test {
-            assert_eq!(m, b"ola");
-
-            send.write_all(b"mundo").await.unwrap();
-            send.finish().await.unwrap();
-        });
-
-        p1_accept.await.unwrap();
-        p2_connect.await.unwrap();
-
-        p2_accept.await.unwrap();
-        p1_connect.await.unwrap();
-
-        cleanup().await;
+
+        for i in 0..TEST_ITERS {
+            println!("round {}", i);
+            let (ep1, ep2, cleanup) = setup_pair().await?;
+
+            let peer_id_1 = ep1.peer_id();
+            eprintln!("peer id 1 {peer_id_1}");
+            let peer_id_2 = ep2.peer_id();
+            eprintln!("peer id 2 {peer_id_2}");
+
+            let endpoint = ep2.clone();
+            let p2_connect = tokio::spawn(async move {
+                let conn = endpoint.connect(peer_id_1, TEST_ALPN, &[]).await?;
+                let (mut send, mut recv) = conn.open_bi().await?;
+                send.write_all(b"hello").await?;
+                send.finish().await?;
+                let m = recv.read_to_end(100).await?;
+                assert_eq!(&m, b"world");
+                anyhow::Ok(())
+            });
+
+            let endpoint = ep1.clone();
+            let p1_accept = tokio::spawn(async move {
+                let conn = endpoint.accept().await.expect("no connection");
+                let (peer_id, alpn, conn) = accept_conn(conn).await?;
+                assert_eq!(peer_id, peer_id_2);
+                assert_eq!(alpn.as_bytes(), TEST_ALPN);
+
+                let (mut send, mut recv) = conn.accept_bi().await?;
+                let m = recv.read_to_end(100).await?;
+                assert_eq!(m, b"hello");
+
+                send.write_all(b"world").await?;
+                send.finish().await?;
+                anyhow::Ok(())
+            });
+
+            let endpoint = ep1.clone();
+            let p1_connect = tokio::spawn(async move {
+                let conn = endpoint.connect(peer_id_2, TEST_ALPN, &[]).await?;
+                let (mut send, mut recv) = conn.open_bi().await?;
+                send.write_all(b"ola").await?;
+                send.finish().await?;
+                let m = recv.read_to_end(100).await?;
+                assert_eq!(&m, b"mundo");
+                anyhow::Ok(())
+            });
+
+            let endpoint = ep2.clone();
+            let p2_accept = tokio::spawn(async move {
+                let conn = endpoint.accept().await.expect("no connection");
+                let (peer_id, alpn, conn) = accept_conn(conn).await?;
+                assert_eq!(peer_id, peer_id_1);
+                assert_eq!(alpn.as_bytes(), TEST_ALPN);
+
+                let (mut send, mut recv) = conn.accept_bi().await?;
+                let m = recv.read_to_end(100).await?;
+                assert_eq!(m, b"ola");
+
+                send.write_all(b"mundo").await?;
+                send.finish().await?;
+                anyhow::Ok(())
+            });
+
+            p1_accept.await??;
+            p2_connect.await??;
```

Comment:
The order between these tasks is not guaranteed.  Therefore it is possible to call `.connect()` before the other has called `.accept()`.  It is not clear to me if this is guaranteed to work if you call connect before accept.

Reply:
I've done some testing and it seems this isn't true.  I can connect without having to call accept first.  (This does make me worry `Conn` is susceptible to some kind of DOS attack)